### PR TITLE
Expand type coverage

### LINKED ISSUE #1081: unitaryHACK - Expand type coverage across Dynamiqs
### Context

Dynamiqs already uses Python typing in parts of the codebase, and type checking is part of the development workflow through `task type`.

Contributors are also expected to type new functions, as described in the [contributing guide](https://github.com/dynamiqs/dynamiqs/blob/main/CONTRIBUTING.md).

However, type coverage is still incomplete. Several important modules are currently excluded from `ty` checking:

```
[tool.ty.src]
exclude = [
    "dynamiqs/integrators",
    "dynamiqs/plot",
    "dynamiqs/qarrays",
    "dynamiqs/random",
    "dynamiqs/utils",
    "dynamiqs/time_qarray.py"
]
```

The current CI does not use the latest version of `ty`, and updating `ty` is known to reveal new typing issues in files that are already type-checked today.

This issue therefore has two equally important parts:

1. make the currently type-checked parts of the library pass with the latest supported version of `ty`;
2. expand type coverage by typing currently excluded files and removing them from the exclusion list.

Improving type coverage will make the codebase easier to maintain, improve IDE/autocomplete support, and make future refactors safer.

### What we want

Update the type-checking setup so Dynamiqs works with the latest supported version of `ty`, then progressively expand type coverage across the library.

The first step is to run `task type` with the latest supported `ty` version and fix the issues reported in files that are already covered by type checking.

The second step is to type currently excluded files or modules. Any file or module that is correctly typed should be removed from the `ty` exclusion list.

The work should follow the existing development workflow:

- add or refine type annotations;
- run `task type`;
- fix reported typing issues;
- run `task all` locally before submitting.

### Important constraints

Typing issues should be fixed properly in the code or annotations.

Do not silence broad classes of errors using global ignores, blanket exclusions, or configuration changes that make `ty` stop checking the relevant code. Suppression comments may be used only when narrowly scoped and clearly justified.

Ignoring typing errors is not a fix. The goal is to increase real type coverage, not to make the checker pass by hiding errors.

### Acceptance criteria / Definition of Done

- The project uses the latest supported version of `ty`.
- The parts of the library that are currently type-checked pass with that version.
- Newly typed files or modules are removed from the `ty` exclusion list.
- All files that are no longer excluded pass type checking.
- Type errors are fixed with real annotations or code changes, not broad ignores or exclusions.
- `task type` passes locally.
- `task all` passes locally.
- The full CI passes.
- Existing behavior is not changed except where needed to fix typing issues.

### Focus

Mostly engineering.

Focus on maintainability, API clarity, developer experience, and static correctness.

### Difficulty

Medium

### PR DESCRIPTION
**Pull request by kwyip**
## Pull Request: Expand type coverage across Dynamiqs

Closes https://github.com/dynamiqs/dynamiqs/issues/1081

## What changed

The pull request pins the type checker, removes a real public module from the exclusion list, and narrows broad exclusions into concrete file-level follow-up work. The current changes move the project in the following direction:

$$
\begin{aligned}
&\text{less excluded code} \\
&+\text{more precise annotations} \\
&+\text{passing } \texttt{task type}\\
 &=\text{safer refactors and better developer experience}.
\end{aligned}
$$

With this pull request, instead of facing a few large directory-level exclusions, Dynamiqs' users now have a file-level map.

### 1. `ty` is pinned to a concrete supported version

The development dependency now uses a concrete `ty` version:

$$
\text{VER}_{\mathrm{ty}} = 0.0.44.
$$

That is great for reproducibility. Everyone running the development workflow is
pointed at the same type checker version rather than floating across whatever
`ty` release happens to be latest on a given day.

In practical terms, this changes the toolchain contract from

$$
\texttt{ty} \to \texttt{ty} = 0.0.44.
$$

### 2. `dynamiqs/random` is now included in type checks

The broad problem called out this excluded module set:

$$
E_{\mathrm{original}} = \{\texttt{integrators}, \texttt{plot},\texttt{qarrays}, \texttt{random}, \texttt{utils},
\texttt{time\\_qarray.py}\}.
$$

The current work removes `dynamiqs/random` from that excluded set:

$$
\texttt{dynamiqs/random} \notin E_{\mathrm{current}}.
$$

Equivalently, the newly covered public module includes

$$
\Delta E_{\mathrm{random}} = \{\texttt{dynamiqs/random}\}.
$$

This can be a win because `dynamiqs/random` is a user-facing utility
surface. It exposes typed random constructors for objects such as real arrays,
complex arrays, Hermitian operators, positive semidefinite matrices, density
matrices, operators, and kets.

The most relevant parameter families are

$$
\texttt{key}:\ \texttt{PRNGKeyArray},
\qquad
\texttt{shape}:\ \texttt{int}\ |\ \texttt{tuple[int, ...]},
$$

and, for Hilbert-space aware constructors,

$$
\texttt{dims}:\ \texttt{int}\ |\ \texttt{tuple[int, ...]},
\qquad
\texttt{batch}:\ \texttt{int}\ |\ \texttt{tuple[int, ...]}.
$$

For composite dimensions $\texttt{dims} = (d_1, \ldots, d_m)$, the Hilbert-space size is of direct rpoduct:

$$
d = \prod_{k=1}^{m} d_k.
$$

That makes the public random-state shape contracts clearer:

$$
\text{ket}(\texttt{key}, \texttt{dims}, \texttt{batch})
\to \texttt{QArray}\ \text{with shape}\ (*\texttt{batch}, d, 1),
$$

$$
\text{dm}(\texttt{key}, \texttt{dims}, \texttt{batch})
\to \texttt{QArray}\ \text{with shape}\ (*\texttt{batch}, d, d).
$$

### 3. Broad exclusions are narrowed to file-level exclusions

Broad directory exclusions can feel like a black box. File-level exclusions are much easier to understand and review.

Before this change, whole directories were excluded:

$$
E_{\mathrm{broad}} = \{\texttt{integrators},\texttt{plot},\texttt{qarrays},\texttt{utils}\}.
$$

After this change, the remaining exclusions are explicit files:

$$
E_{\mathrm{current}} = \{f_1, f_2, \ldots, f_n\},
\qquad f_i = \text{one known file-level typing task}.
$$

This is a NICE migration shape because it increases real coverage without
pretending the remaining work is done. It also gives future users a much
clearer checklist:

$$
\text{pick } f_i
\quad \to \quad
\text{fix annotations or code}
\quad \to \quad
\text{remove } f_i \text{ from } E_{\mathrm{current}}.
$$

Newly visible package entry points and helper files include examples like

`dynamiqs/plot/colormaps.py`
`dynamiqs/qarrays/layout.py`
`dynamiqs/utils/global_settings.py`
`dynamiqs/utils/optimal_control.py`

The pull request changes the type coverage map from broad silencing to actionable work.

### Type coverage utilities notebook

The notebook `type_coverage_example/type_coverage_utilities.ipynb` is a user-friendly
example of how to write typed Dynamiqs utilities. 

The central pattern is

$$
\text{broad public input}
\xrightarrow{\ \text{convert once at the boundary}\ }
\text{stable internal type}
\xrightarrow{\ \text{return explicitly}\ }
\texttt{QArray}\ \text{or}\ \texttt{Array}.
$$

Concretely:

- use `QArrayLike` for quantum objects accepted from users;
- use `ArrayLike` for numerical arrays accepted from users;
- convert with `dq.asqarray()` or `jnp.asarray()`;
- return `QArray` or `jax.Array`;
- use `typing.assert_type()` to document and check the intended static contract.
- verify `ty --version` and run `task type`

#### `normalized_state(state)`

This utility accepts a friendly qarray-like input and returns a normalized
Dynamiqs `QArray`.

Relevant parameter:

$$
\texttt{state}:\ \texttt{QArrayLike}.
$$

Conversion and return contract:

$$
\texttt{QArrayLike}
\xrightarrow{\ \texttt{dq.asqarray}\ }
\texttt{QArray}
\xrightarrow{\ \texttt{unit}\ }
\texttt{QArray}.
$$

Function type:

$$
\text{normalized\\_state}:
\texttt{QArrayLike} \to \texttt{QArray}.
$$

It preserves Dynamiqs ergonomics at the call site while giving the implementation a precise internal type.

#### `centered_grid(points)`

This utility accepts Python, NumPy, or JAX-style numerical input and returns a
JAX `Array`.

Relevant parameter:

$$
\texttt{points}:\ \texttt{ArrayLike}.
$$

For input values $x_1, \ldots, x_N$, it computes

$$
\bar{x} = \frac{1}{N}\sum_{j=1}^{N} x_j,
\qquad
\tilde{x}_i = x_i - \bar{x}.
$$

Function type:

$$
\text{centered\\_grid}:
\texttt{ArrayLike} \to \texttt{Array}.
$$

It demonstrates the array-side mirror of the qarray boundary conversion pattern, using `jnp.asarray()` as the normalization step.

#### `random_density_matrix(key, dims)`

This utility wraps `dq.random.dm()` with a precise return annotation.

Relevant parameters:

$$
\texttt{key}:\ \texttt{PRNGKeyArray},
\qquad
\texttt{dims}:\ \texttt{int}\ |\ \texttt{tuple[int, ...]}.
$$

For $\texttt{dims} = (d_1, \ldots, d_m)$,

$$
d = \prod_{k=1}^{m} d_k,
\qquad
\rho \in \mathbb{C}^{d \times d}.
$$

Function type:

$$
\text{random\\_density\\_matrix}:
(\texttt{PRNGKeyArray},\ \texttt{int}\ |\ \texttt{tuple[int, ...]})
\to \texttt{QArray}.
$$

It connects the newly covered `dynamiqs/random` module directly to user-facing typed helper code.

#### `real_expectations(observables, state)`

This utility accepts qarray-like observables and a qarray-like state, then
returns real expectation values as a JAX `Array`.

Relevant parameters:

$$
\texttt{observables}:\ \texttt{Iterable[QArrayLike]},
\qquad
\texttt{state}:\ \texttt{QArrayLike}.
$$

For each observable $O_k$, the expectation value is

$$
\langle O_k \rangle =
\begin{cases}
\langle \psi | O_k | \psi \rangle, & \text{if the state is a ket } \psi,\\
\text{Tr}(O_k\rho), & \text{if the state is a density matrix } \rho.
\end{cases}
$$

The utility returns the real component:

$$
y_k = \text{Re}\langle O_k \rangle.
$$

Function type:

$$
\text{real\\_expectations}:
(\texttt{Iterable[QArrayLike]},\ \texttt{QArrayLike})
\to \texttt{Array}.
$$

It essentially turns a common scientific pattern into a statically clear helper without taking flexibility away from users.

#### `RandomStateExperiment`

This dataclass shows how a notebook experiment can grow into structured typed
code.

Relevant fields:

$$
\texttt{dims}:\ \texttt{int}\ |\ \texttt{tuple[int, ...]},
\qquad
\texttt{sample\\_points}:\ \texttt{ArrayLike}.
$$

Relevant methods:

$$
\text{state}(\texttt{key}:\ \texttt{PRNGKeyArray})
\to \texttt{QArray},
$$

$$
\text{density\\_matrix}(\texttt{key}:\ \texttt{PRNGKeyArray})
\to \texttt{QArray},
$$

$$
\text{centered\\_points}() \to \texttt{Array}.
$$

It gives users a realistic pattern for carrying type clarity from exploratory code into reusable helpers.

####  Verify `ty --version` and run `task type`

Should output `ty 0.0.44` and 
```
>>> ty check dynamiqs All checks passed!
```

### Demo

https://github.com/user-attachments/assets/1bbf0b23-f497-4159-9c96-e1f4fb9a178f

### Generative AI/LLM disclosure

Code and Logic Architecture/Design:

- [ ] The code and logic architecture/design contain no generative AI/LLM
- [x] The code and logic architecture/design are partially performed by generative AI/LLM
  
  - **99%** performed by the author(s)
- [ ] The code and logic architecture/design are fully performed by generative AI/LLM

Code content:

- [ ] The code contains no generative AI/LLM work
- [x] The code is partially written by generative AI/L
  
  - **69%** performed by the author(s)
- [ ] The code is fully written by generative AI/LLM

Code review:

- [x] The code review contains no generative AI/LLM
- [ ] The code review is partially done by generative AI/LLM
  
  - **X%** performed by the author(s)
- [ ] The code review is fully done by generative AI/LLM

Code tests:

- [x] The code tests contain no generative AI/LLM
- [ ] The code tests are partially written by generative AI/LLM
  
  - **X%** performed by the author(s)
- [ ] The code tests are fully written by generative AI/LLM

### CODE REVIEW

**Comment by Nicolas-Lepage** [comment]
Thanks for the PR! Unfortunately, we ended up merging another contribution that addressed the same issue, so I'll be closing this PR. Thanks again for taking the time to work on it.
